store/iavl/proof.go

```diff
@@ -0,0 +1,122 @@
+// NOTE: CommitmentProof currently can either be ExistenceProof or NonexistenceProof
+type CommitmentOp struct {
+	Key   []byte
+	Proof *ics23.CommitmentProof
+}
+
+var _ merkle.ProofOperator = CommitmentOp{}
+
+func NewCommitmentOp(key []byte, proof *ics23.CommitmentProof) CommitmentOp {
+	return CommitmentOp{
+		Key:   key,
+		Proof: proof,
+	}
+}
+
+// CommitmentOpDecoder takes a merkle.ProofOp and attempt to decode it into a CommitmentOp ProofOperator
+// The proofOp.Data is just a marshalled CommitmentProof. The Key of the CommitmentOp is extracted
+// from the unmarshalled proof
+func CommitmentOpDecoder(pop merkle.ProofOp) (merkle.ProofOperator, error) {
+	if pop.Type != ProofOpIAVLCommitment {
+		return nil, fmt.Errorf("unexpected ProofOp.Type; got %v, want %v", pop.Type, ProofOpIAVLCommitment)
+	}
+	var op CommitmentOp
+	proof := &ics23.CommitmentProof{}
+	err := proof.Unmarshal(pop.Data)
+	if err != nil {
+		return nil, err
+	}
+	op.Proof = proof
+
+	// Get Key from proof for now
+	if existProof, ok := op.Proof.Proof.(*ics23.CommitmentProof_Exist); ok {
+		op.Key = existProof.Exist.Key
+	} else if nonexistProof, ok := op.Proof.Proof.(*ics23.CommitmentProof_Nonexist); ok {
+		op.Key = nonexistProof.Nonexist.Key
+	} else {
+		return nil, errors.New("proof type unsupported")
+	}
+	return op, nil
+}
+
+func (op CommitmentOp) GetKey() []byte {
+	return op.Key
+}
+
+func (op CommitmentOp) Run(args [][]byte) ([][]byte, error) {
+	// Only support an existence proof or nonexistence proof (batch proofs currently unsupported)
+	switch len(args) {
+	case 0:
+		// Args are nil, so we verify the absence of the key.
+		nonexistProof, ok := op.Proof.Proof.(*ics23.CommitmentProof_Nonexist)
+		if !ok {
+			return nil, errors.New("proof is not a nonexistence proof and args is nil")
+		}
+
+		// get root from either left or right existence proof. Note they must have the same root if both exist
+		// and at least one proof must be non-nil
+		root, err := nonexistProof.Nonexist.Left.Calculate()
+		if err != nil {
+			// Left proof may be nil, check right proof
+			root, err = nonexistProof.Nonexist.Right.Calculate()
+			if err != nil {
+				return nil, errors.New("could not calculate root from nonexistence proof")
+			}
+		}
+
+		absent := ics23.VerifyNonMembership(ics23.IavlSpec, root, op.Proof, op.Key)
+		if !absent {
+			return nil, fmt.Errorf("proof did not verify absence of key: %s", string(op.Key))
+		}
+
+		return [][]byte{root}, nil
+
+	case 1:
+		// Args is length 1, verify existence of key with value args[0]
+		existProof, ok := op.Proof.Proof.(*ics23.CommitmentProof_Exist)
+		if !ok {
+			return nil, errors.New("proof is not a existence proof and args is length 1")
+		}
+		// For subtree verification, we simply calculate the root from the proof and use it to prove
+		// against the value
+		root, err := existProof.Exist.Calculate()
+		if err != nil {
+			return nil, errors.New("could not calculate root from existence proof")
```

Comment:
I also wonder about wrapping the errors.
I thought `types/errors` aka `sdkerrors` was the norm for error reporting. This would give a stack trace, and abci return code, and the chance to grab info from the original error.